### ISSUE TITLE
Does DeepAR needs features and target to be normalized

### ISSUE THREAD

**Issue by ahmedmeer** [awslabs/amazon-sagemaker-examples#1312]
I am predicting sales (say dollars, which is the target) using DeepAR by training multiple related time series. Following are details of features and target (value ranges):

Target variable: (sales in dollars) typical values range from 3000 to 40000 (avg: 22000)

Dynamic Feature 1: values range from 1 to 50 (int)
Dynamic Feature 2: values range from 1 to 10 (int)
Dynamic Feature 3: possible values are 0 and 1 (int/binary)
Dynamic Feature 4: values range from 1 to 5 (int)
Dynamic Feature 5: values range from 0 to 1 (float)

following are cat features' values after encoding them to integer from string values,

Categorical Feature 1: values range from 0 to 2
Categorical Feature 1: values range from 0 to 17
Categorical Feature 1: values range from 0 to 153
Categorical Feature 1: values range from 0 to 1378
Categorical Feature 1: values range from 0 to 44
Categorical Feature 1: values range from 0 to 2

Considering above value ranges for my features and target, is it recommended to use any kind of data normalization/scaling or this is something that DeepAR manages on its own. If yes, which normalization technique works best with DeepAR and will it be required for both features and target (to bring them both on same scale) or only on the features. Thanks in advance.

**Comment by songzhm** [comment]
Hi @nxr1691 , I've been recently learning and trying DeepAR as well. I don't know the exactly answer as I didn't know the details of the implementation. However, here is what I learnt from the [paper](https://arxiv.org/abs/1704.04110) and what I observed from my experiments:

From the DeepAR paper, it is claimed that there is scale-handling process built into the algorithm, more details can be found in section 3.3 of the paper. In addition, from the paper section 3.4, the authors said that
> We standardize all covariates to have zero mean and unit variance.
But I don't know does this refers to something builtin to the algorithm, or this is addition process they took for empirical study.

Based on my experience, it seems okay to feed different magnitude levels of time series directly to DeepAR, and the output would match the origin time series pretty well. 

However, there are some caveats need pay attention: 

1. When I tried to feed some dynamic features time series with significantly different level of magnitudes, the performance dropped a lot, and I found that it is very hard for the training to converge. And most of the predictions are constant across the forecasting horizon.

2. The performance of the model got significantly improved after I scaled (MinMaxScaler) the dynamic features to a similar magnitude range compare to target time series. And the training loss curve looks much normal, and the predictions are much more reasonable.

In addition, my intuition is that DL models is trained well with standardized/normalized data, and DeepAR use RNN layers. Hence, no matter if DeepAR has builtin process to scale/normalize input, if we can preprocess the input before feeding into DeepAR, it won't be harmful for the performance, at least, if not able to boost performance.

Those are what my thoughts are based on my current understanding and experiments from the past week's experience with DeepAR. It might contain premature thoughts. Looking forward to more insights from experts.

**Comment by nxr1691** [comment]
Hi @songzhm , thanks for sharing your thoughts in detail. This seems interesting. Just a quick question, in your 2nd bullet point, did you scale your dynamic features to meet the scale of your target time series ? or did you just scale them to make all dynamic features in the same range ? Because the former case might not be possible in my scenario as my target time series is sales in dollars whose values can really fluctuate with thousands of magnitude.

**Comment by songzhm** [comment]
No, I did scale the dynamic features to a certain range to a level that is more comparable to my target time series, but not necessary match exotically the magnitude level of my target time series. For example, some of target time series usually ranging from 0-30, some target series is ranging 40-60, etc.. Then I scale my dynamic feature to a range roughly 0-100.

I don't know if 0-1, instead of 0-100, will better help for the training, but I did see the convergence is better in my training process by using 0-100.

Let me know if this works for you or if there is any thing that I might misunderstood.

**Comment by nxr1691** [comment]
Got it. When you say the convergence is better, does that mean you got better accuracy on your validation set/unseen data after scaling ? Or does that mean better in terms of less time consumed to converge (but accuracy was more or less same)

**Comment by jcabralc** [comment]
@songzhm when you normalized your dynamic features you did normalize one by one individually?

I have the same problem, but my dynamic feature differ in a very high range between them, for example, I have one in 1,34-3,248 range, and other with 29.862-233.735

my target is between 1-32629

**Comment by Sandy4321** [comment]
yes it is not clear what dynamic feature is?
where is no description 
https://docs.aws.amazon.com/sagemaker/latest/dg/deepar-in-formats.html
and in paper not mentioned ?
https://arxiv.org/pdf/1704.04110.pdf

if somebody can share link to explanation what is it?

**Comment by nxr1691** [comment]
@Sandy4321
Dynamic features are related time series that correlate with your target time series. For example, if your target time series is sales on daily level, one of the dynamic features could be the day of year, another dynamic feature could be the daily stock price etc.

Dynamic and Cat features are explained in the following link
https://docs.aws.amazon.com/sagemaker/latest/dg/deepar.html#deepar-inputoutput

**Comment by Sandy4321** [comment]
great thanks
do you have detailed algorithm description?
since referenced paper
https://arxiv.org/pdf/1704.04110.pdf
Submitted on 13 Apr 2017 (v1)
 does not have details
is this paper still relevant or you added new algorithms?
for example categorical values mentioned only once
''
Further, we include a single categorical item
feature, for which an embedding is learned by the model. In the retail demand forecasting data sets,
the item feature corresponds to a (coarse) product category (e.g. “clothing”), while in the smaller
data sets it corresponds to the item’s identity, allowing the model to learn item-specific behavior.
''

**Comment by nxr1691** [comment]
The algorithm details are mentioned in the paper you referred. I didn't find any deeper explanation anywhere else in the aws docs. The aws docs mainly focus on how to use the model and prepare your training data and features etc. I am following the aws docs and the sample notebooks they have shared for DeepAR.

Regarding categorical item, yes you are right, it is supposed to differentiate between different time series. If we have a problem where there is only one category, the categorical features might not make much sense. The essence of DeepAR lies in using multiple time series that are related to each other and can be differentiated by categorical features.

**Comment by Sandy4321** [comment]
https://github.com/brunoklein99/deepar
they did not implemented categorical values , may be by the same reason - aws keep it in secret how to do this  
What this implementation does NOT contain
Two significant pieces are left out at this time, albeit trivial to implement.

The joint embedding learning for item categorization

**Comment by JingyiWei8** [comment]
Hey all,

I might be a bit late for the discussion. But I am facing the same question with my data. According to this video from AWS (23:05 mins) https://www.youtube.com/watch?v=g8UYGh0tlK0 ,  it seems it's required. But other sources from AWS indicates otherwise, like what `songzhm` mentioned. What do you guys found in your cases? Does DeepAR predict better with scaling? If so, what approach did you guys used?
